[WIP] Structure revamp of Serverless docs

abrennan89 opened the pull request:
Major structural changes.

1. Improved getting started intro content. Next steps changed to additional resources.
1. New getting_started folder; installation folder removed, topics now split between main serverless root folder and getting_started.
1. Edited topicmap to reflect moved topics.
1. Edits / updates to removing serverless section (later todo: split last part of this into its own module re deleting CRDs).
1. Updated xrefs for moved files.
1. Improvements to content based on last minimalism review.
1. Update code blocks to use new formal, [source,x] etc.
1. Moved subscription module content to assembly, there were xrefs in the module for some reason.
1. Sinkbinding moved to event_sources
1. Metering, tracing and HA moved to knative_serving
1. supported config line with link got its own module so that it can be reused
1. moved arch sections into getting started before sections for installing each component respectively
1. New about serverless section; previously removed due to minimalism guidelines re using this as a heading, but it's used a lot in other places, so re-adding it until there's a viable alt

To do:

- [x] Follow up with sheet updates to have links rerouted to new locations.
- [x] CP to all required branches and resolve any conflicts.
  
  - #25221 CP to 4.3
  - #25262 CP to 4.4
  - #25263 CP to 4.5
  - #25264 CP to 4.6

vikram-redhat commented:
@abrennan89 ping me when this merges.

cardil reviewed (approved):
/lgtm

From QE. I didn't saw any obvious problems.

bergerhoffer commented on _topic_map.yml:
```diff
@@ -2407,12 +2393,21 @@ Topics:
   - Name: Using NVIDIA GPU resources with serverless applications
     File: gpu-resources
+# Release notes
+- Name: Release Notes
```
I know it was like this before it was moved, but "Release Notes" really should be "Release notes". Same w/ the name of the assembly.

bergerhoffer commented on serverless/knative_eventing/serverless-kn-trigger.adoc:
```diff
@@ -65,8 +70,10 @@ include::modules/kn-trigger-describe.adoc[leveloffset=+1]
 
 To delete a trigger, enter the following command:
+
+[source,terminal]
```
I see you adding [source,terminal] a few places, but in looking through the entire serverless book, there are a ton that don't have this yet. I might suggest to not worry about that in this PR, and then open a PR after this is merged to add those in.

abrennan89 replied:
Cool, I'll leave the ones I've added in this PR as are though if that's OK? :slightly_smiling_face: Or do I need to revert them?

bergerhoffer commented on serverless/knative_eventing/serverless-subscriptions.adoc:
```diff
@@ -6,6 +6,57 @@ include::modules/common-attributes.adoc[]
-include::modules/serverless-creating-subscriptions.adoc[leveloffset=+1]
+[IMPORTANT]
+====
+Knative Eventing is a Technology Preview feature. The InMemoryChannel type is provided for development use only, and should not be used in a production environment.
```
We have a `technology-preview.adoc` file that we include so that we use consistent wording for TP features. Suggest following that here (just search the repo for that file and you'll see how it's used).

abrennan89 replied:
I actually use it somewhere else, but this note is specifically to call out the thing about InMemoryChannel rather than the TP status. I'll remove the first line to make this the focus. Thanks.

bergerhoffer commented on modules/serverless-supported-configs.adoc:
```diff
@@ -0,0 +1,4 @@
+[id="serverless-supported-configs_{context}"]
```
This file is missing the metadata that says which assemblies it's used in.

abrennan89 replied:
Oops, added thanks.

bergerhoffer commented on modules/serverless-supported-configs.adoc:
```diff
@@ -0,0 +1,4 @@
+[id="serverless-supported-configs_{context}"]
+= Supported Configurations
```
Use sentence case

bergerhoffer commented on modules/serverless-supported-configs.adoc:
```diff
@@ -0,0 +1,4 @@
+[id="serverless-supported-configs_{context}"]
+= Supported Configurations
+
+The set of supported features, configurations, and integrations for {ServerlessProductName}, current and past versions, are available at the link:https://access.redhat.com/articles/4912821[Supported Configurations page].
```
could you simplify to 

```
The set of supported features, configurations, and integrations for all versions of {ServerlessProductName} are available at the link:https://access.redhat.com/articles/4912821[Supported Configurations page].
```
Also maybe "at the" -> "on the" ?

bergerhoffer commented on serverless/about-serverless.adoc:
```diff
@@ -0,0 +1,78 @@
+
+include::modules/serverless-supported-configs.adoc[leveloffset=+1]
+
+[id="about-serverless-serving"]
```
Suggest making all these subsections into modules and then including them from the assembly.

bergerhoffer commented on serverless/about-serverless.adoc:
```diff
@@ -0,0 +1,78 @@
+
+toc::[]
+
+{ServerlessProductName} enables developers to deploy and manage cloud native applications that can run anywhere, without provisioning or managing servers.
```
cloud native -> cloud-native?

bergerhoffer commented on serverless/about-serverless.adoc:
```diff
@@ -0,0 +1,78 @@
+{ServerlessProductName} enables developers to deploy and manage cloud native applications that can run anywhere, without provisioning or managing servers.
+These routine tasks are abstracted away by the platform, allowing developers to push code to production much faster than in traditional models.
+
+{ServerlessProductName} is based on the open source link:https://knative.dev/[Knative] project, which provides portability and consistency across hybrid and multi-cloud environments by enabling an enterprise-grade serverless platform.
```
The way this is worded, I'm not sure whether the second part is saying that Knative project enables an enterprise-grade platform, or if it's supposed to mean that OpenShift Serverless does.

bergerhoffer commented on serverless/about-serverless.adoc:
```diff
@@ -0,0 +1,78 @@
+{ServerlessProductName} is based on the open source link:https://knative.dev/[Knative] project, which provides portability and consistency across hybrid and multi-cloud environments by enabling an enterprise-grade serverless platform.
+
+Developers can use a single platform for hosting their microservices, legacy, and serverless applications, and can use the provided Kubernetes native APIs, as well as familiar languages and frameworks, to deploy applications and stateless container workloads.
+This improves developer productivity by removing the need for complex infrastructure set up before apps can be deployed.
```
Should this say "the need for complex infrastructure to be set up before..." ?

Or maybe "the need for a complex infrastructure setup before..." ?

bergerhoffer commented on serverless/about-serverless.adoc:
```diff
@@ -0,0 +1,78 @@
+Developers can use a single platform for hosting their microservices, legacy, and serverless applications, and can use the provided Kubernetes native APIs, as well as familiar languages and frameworks, to deploy applications and stateless container workloads.
+This improves developer productivity by removing the need for complex infrastructure set up before apps can be deployed.
+
+Operational costs can be reduced by implementing out of the box {ServerlessProductName} solutions for automating operational tasks, such as scaling apps, by installing the {ServerlessProductName} Knative Serving and Eventing components.
```
out-of-the-box

bergerhoffer replied:
Also this sentence is a little long/complex. I'd consider breaking it up a little if possible.

bergerhoffer commented on serverless/about-serverless.adoc:
```diff
@@ -0,0 +1,78 @@
+=== Knative Serving CRDs
+// TODO: Request a good Serving arch diagram
+
+Service:: The `service.serving.knative.dev` CRD automatically manages the life cycle of your workload to ensure that the application is deployed and reachable through the network. It creates a Route, a Configuration, and a new Revision for each change to a user created Service, or custom resource. Most developer interactions in Knative are carried out by modifying Services.
```
* Should "Revision" and "Configuration" be lowercase? (Same elsewhere in the PR)
* user created -> user-created?
* Probably lose the comma before custom resource

abrennan89 replied:
I don't think so re lowercase - It's kind of like service vs. Service, these are referring to a specific type of object as opposed to the general concept of "configurations" or "revisions" etc, so it makes sense to me to give them caps?

This should probably be part of the overall discussions happening around Kubernetes objects, what to capitalize etc, so any guidance is appreciated, I'm sort of waiting to see what the OpenShift guidelines end up being before I recommend anything specific for Serverless or Knative :slightly_smiling_face: 

Maybe a follow up issue/PR to track adding these specific resource types to the glossary would be worthwhile?

bergerhoffer reviewed (commented):
I didn't finish a whole review, but sharing the feedback so far

bobfuru commented on serverless/knative_eventing/serverless-kn-trigger.adoc:
```diff
@@ -18,7 +18,7 @@ Before you can use triggers, you will need:
 * An available broker, either the `default` broker or one that you have created.
 +
-You can create the `default` broker either by following the instructions on xref:../serverless-using-brokers.adoc#serverless-using-brokers[Using brokers with Knative Eventing], or by using the `--inject-broker` flag while creating a trigger. Use of this flag is described in the procedure below.
+You can create the `default` broker either by following the instructions on xref:../../serverless/knative_eventing/serverless-using-brokers.adoc#serverless-using-brokers[Using brokers with Knative Eventing], or by using the `--inject-broker` flag while creating a trigger. Use of this flag is described in the procedure below.
```
s/procedure below/following procedure/
**below** _adj_
Do not use to indicate a relative location in a document, as in “the information below.”

bobfuru commented on serverless/getting_started/installing-openshift-serverless.adoc:
```diff
@@ -6,47 +6,53 @@ include::modules/common-attributes.adoc[]
-For more information on using the MachineSet API, see the documentation on xref:../../machine_management/creating_machinesets/creating-machineset-aws.adoc#creating-machineset-aws[Creating MachineSets].
-
-For more information on scaling a MachineSet manually, see the documentation on xref:../../machine_management/manually-scaling-machineset.adoc#manually-scaling-machineset[manually scaling MachineSets].
+The minimum requirement to use {ServerlessProductName} is an {product-title} cluster with 10 CPUs and 40GB memory.
```
s/40GB/40 GB/

bobfuru commented on serverless/getting_started/installing-openshift-serverless.adoc:
```diff
@@ -6,47 +6,53 @@ include::modules/common-attributes.adoc[]
+[id="installing-openshift-serverless-cluster-sizing-reqs-advanced"]
+=== Additional cluster sizing requirements for advanced use cases
 
 For more advanced use-cases such as logging or metering on {product-title}, you must deploy more resources. Recommended requirements for such use-cases are 24 CPUs and 96GB of memory.
```
s/96GB/96 GB/
For consistency with this doc and ISG, could you change this also?

bobfuru commented on serverless/getting_started/installing-openshift-serverless.adoc:
```diff
@@ -6,47 +6,53 @@ include::modules/common-attributes.adoc[]
+[id="installing-openshift-serverless-ha"]
+==== Requirements for high availability
+
+If you have xref:../../serverless/knative_serving/serverless-HA.adoc#serverless-HA[high availability] (HA) enabled on your cluster, this requires between 0.5 - 1.5 cores and between 200MB - 2GB of memory for each replica of the Knative Serving control plane.
```
s/200MB - 2GB/200 MB - 2 GB/

bobfuru commented on serverless/about-serverless.adoc:
```diff
@@ -0,0 +1,78 @@
+This improves developer productivity by removing the need for complex infrastructure set up before apps can be deployed.
+
+Operational costs can be reduced by implementing out of the box {ServerlessProductName} solutions for automating operational tasks, such as scaling apps, by installing the {ServerlessProductName} Knative Serving and Eventing components.
+You can fine tune system and application settings by creating or modifying Kubernetes custom resources.
```
s/fine tune/fine-tune/

bobfuru commented on serverless/getting_started/installing-knative-serving.adoc:
```diff
@@ -10,7 +10,7 @@ After you install the {ServerlessOperatorName}, you can install Knative Serving
 This guide provides information about installing Knative Serving using the default settings. However, you can configure more advanced settings in the KnativeServing custom resource definition.
 
-For more information about configuration options for the KnativeServing custom resource definition, see xref:../installing_serverless/serverless-install-config-options.adoc#serverless-install-config-options[Advanced installation configuration options].
+For more information about configuration options for the KnativeServing custom resource definition, see xref:../../serverless/serverless-install-config-options.adoc#serverless-install-config-options[Advanced installation configuration options].
```
You have "Custom Resource Definition (CRD)" elsewhere in this PR, so I'd capitalize here for consistency.

abrennan89 replied:
I think it should actually be `CustomResourceDefinition` according to https://github.com/openshift/openshift-docs/pull/22440 - I forgot about this tbh.

It doesn't seem to be consistently applied in a bunch of places across docs yet though, most seem to use "Custom Resource Definition (CRD)" or similar, so I will use this for now and put in a separate PR later to update all the instances to `CustomResourceDefinition` if that's OK?

bobfuru commented on serverless/getting_started/installing-openshift-serverless.adoc:
```diff
@@ -6,47 +6,53 @@ include::modules/common-attributes.adoc[]
 toc::[]
 
-This guide walks cluster administrators through installing the {ServerlessOperatorName} to an {product-title} cluster.
+This guide provides information about supported configurations, prerequisites and installation procedures for {ServerlessProductName}.
```
s/prerequisites and installation/prerequisites, and installation/

bobfuru commented on serverless/knative_eventing/serverless-subscriptions.adoc:
```diff
@@ -6,6 +6,57 @@ include::modules/common-attributes.adoc[]
+.Prerequisites
+
+* You must have a current installation of xref:../../serverless/getting_started/installing-openshift-serverless.adoc#serverless-install-web-console_installing-openshift-serverless[{ServerlessProductName}], including Knative Serving and Eventing, in your {product-title} cluster. This can be installed by a cluster administrator.
+* If you do not have an existing sink that you wish to use, create a Service to use as a sink by following the documentation on xref:../../serverless/getting_started/serving-creating-managing-apps.adoc#serving-creating-managing-apps[Creating and managing serverless applications].
```
s/wish/want/

bobfuru commented on serverless/knative_eventing/serverless-subscriptions.adoc:
```diff
@@ -6,6 +6,57 @@ include::modules/common-attributes.adoc[]
+
+.Procedure
+
+. Create a Subscription object to connect a channel to a service, by creating a YAML file containing the following:
```
s/service, by/service by/

bobfuru commented on serverless/knative_eventing/serverless-subscriptions.adoc:
```diff
@@ -6,6 +6,57 @@ include::modules/common-attributes.adoc[]
++
+<1> Name of the subscription.
+<2> Configuration settings for the channel that the subscription connects to.
+<3> Configuration settings for event delivery. This tells the subscription what happens to events that cannot be delivered to the subscriber. When this is configured, events that failed to be consumed are sent to the `deadLetterSink`. The event is dropped, no re-delivery of the event is attempted, and an error is logged in the system. The `deadLetterSink` value must be a link:https://pkg.go.dev/k ...
```
Not sure if this would need an ack from Product Experience/Support because it is an external link?

abrennan89 replied:
Removed for now.

bobfuru commented on serverless/knative_serving/serverless-metering.adoc:
```diff
@@ -8,11 +8,11 @@ toc::[]
 As a cluster administrator, you can use metering to analyze what is happening in your {ServerlessProductName} cluster.
 
-For more information about metering on {product-title}, see xref:../metering/metering-about-metering.adoc#about-metering[About metering].
+For more information about metering on {product-title}, see xref:../../metering/metering-about-metering.adoc#about-metering[About metering].
```
To me, "For more information about metering on OpenShift" sounds odd. Maybe "...about {product-title} metering" or "about using metering on {product-title}..."?

bobfuru commented on serverless/serverless-support.adoc:
```diff
@@ -8,7 +8,7 @@ toc::[]
 include::modules/serverless-getting-support.adoc[leveloffset=+1]
 
-[id="serverless-support-gather-info_{context}"]
+[id="serverless-support-gather-info"]
```
Just curious why you removed the `_{context}` separator here? I see the variable is still defined in the assembly (line 4). (I confess that I don't fully understand how we use this variable.)

abrennan89 replied:
Afaik it's just needed in modules, not assembly ID headings, but I could be wrong?

vikram-redhat replied:
The templates for mod docs don't specify the need to have an anchor for headings in assemblies. @openshift/team-documentation - do we need to specify this?

bergerhoffer replied:
Nope, the _{context} part should only be needed for modules.

bobfuru reviewed (commented):
Overall, LGTM! I added a few minor suggestions in addition to what Andrea already added earlier in her review.

abrennan89 commented:
Added changes from reviews except for the couple of comments I've left @bobfuru @bergerhoffer - if there's no other feedback this should be OK to merge.

I've added links in the description to the PRs for CPing all these changes to 4.3 - 4.6 branches and dealing with any conflicts.
@cardil these other PRs should be fine to QE review once we have confirmation that there are no more peer review changes.

Thanks folks!

bobfuru reviewed (commented):
Updates LGTM - please squash commits. And if possible, let's wait until @bergerhoffer can also re-review. Thanks!

vikram-redhat commented:
Hey @abrennan89 - does this restructure undo any of the changes that we recently did for which we added some redirects?

bergerhoffer commented on serverless/serverless-release-notes.adoc:
```diff
@@ -1,12 +1,12 @@
 include::modules/serverless-document-attributes.adoc[]
 [id="serverless-release-notes"]
-= {ServerlessProductName} Release Notes
+= {ServerlessProductName} Release notes
```
Sorry I mislead you with that other comment, this should technically be "release notes" both lowered.

bergerhoffer commented:
@abrennan89 I re-reviewed and had 2 comments that might have been collapsed that you didn't ack. It's fine if you chose to not doing anything for those, but just wanted to make sure you didn't miss them just because they were collapsed.

I had just one last thing, and then other than @vikram-redhat's question and needing to squash, the rest LGTM.
